src/v/cluster_link/topic_reconciler.cc

```diff
@@ -0,0 +1,326 @@
+/**
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/dev/licenses/rcl.md
+ *
+ */
+
+#include "cluster_link/topic_reconciler.h"
+
+#include "cluster/types.h"
+#include "cluster_link/logger.h"
+#include "cluster_link/model/types.h"
+#include "cluster_link/utils/topic_properties_utils.h"
+#include "model/fundamental.h"
+#include "ssx/future-util.h"
+
+#include <exception>
+
+using namespace std::chrono_literals;
+
+namespace cluster_link {
+topic_reconciler::topic_reconciler(
+  kafka::data::rpc::topic_creator* topic_creator,
+  kafka::data::rpc::topic_metadata_cache* topic_metadata_cache,
+  link_registry* link_registry,
+  ss::lowres_clock::duration run_interval)
+  : _topic_creator(topic_creator)
+  , _topic_metadata_cache(topic_metadata_cache)
+  , _link_registry(link_registry)
+  , _run_interval(run_interval) {}
+
+ss::future<> topic_reconciler::start() {
+    vlog(cllog.info, "Starting topic reconciler");
+    _reconciler_timer.set_callback([this] {
+        ssx::spawn_with_gate(_gate, [this] {
+            return execute()
+              .handle_exception([](const std::exception_ptr& e) {
+                  vlog(cllog.error, "Topic reconciler failed: {}", e);
+              })
+              .finally([this] {
+                  if (_as.abort_requested()) {
+                      return;
+                  }
+                  _reconciler_timer.arm(_run_interval);
+              });
+        });
+    });
+    _reconciler_timer.arm(_run_interval);
+    co_return;
+}
+
+ss::future<> topic_reconciler::stop() noexcept {
+    vlog(cllog.info, "Stopping topic reconciler");
+    _reconciler_timer.cancel();
+    _as.request_abort();
+    co_await _gate.close();
+    vlog(cllog.info, "Topic Reconciler stopped");
+}
+
+void topic_reconciler::trigger(model::id_t link_id) {
+    vlog(cllog.trace, "Triggering topic reconciler for link id={}", link_id);
+    ssx::spawn_with_gate(_gate, [this, link_id] {
+        return execute(link_id).handle_exception(
+          [](const std::exception_ptr& e) {
+              vlog(cllog.error, "Topic reconciler failed: {}", e);
+          });
+    });
+}
+
+ss::future<> topic_reconciler::execute(std::optional<model::id_t> link_id) {
+    vlog(cllog.trace, "Executing topic reconciler, link_id: {}", link_id);
+    try {
+        auto units = co_await _reconciler_mutex.get_units(_as);
+        if (link_id.has_value()) {
+            co_return co_await do_reconcile_topic(link_id.value());
+        }
+
+        auto link_ids = _link_registry->get_all_link_ids();
+        co_return co_await ss::parallel_for_each(
+          link_ids, [this](model::id_t id) { return do_reconcile_topic(id); });
+    } catch (const std::exception& e) {
```

Comment:
do we need to catch the exception here ?
It seems that we are trying to handle exception in the higher level, so either do it here or there